Improved Camera2D limits handling when limits are smaller than screen rect

ogapo opened the pull request:
This commit improves the behavior of 2D camera limits for edge cases when the limits are smaller than the screen rect.

Presently the RIGHT and TOP limit take precedence simply because they happened to be coded to apply second. This effectively means the part of the scene within the desired limit bounds (typically the focus) is placed in the top right corner of the viewport.
This change adjusts behavior to split the difference in cases where both the LEFT/RIGHT or TOP/BOTTOM limits both apply  resulting in the target area being placed in the center of the viewport which is more natural and expected. 

There's no change in the typical case where the camera limits are larger than the screen rect.

Adjusted the if-else chain so there's no performance cost for this change (same number of tests are performed).

KoBeWi commented:
You included unrelated commit and there are still conflicts.

ogapo commented:
> You included unrelated commit and there are still conflicts.

Sorry I don't actually know why it requested review. Seems to have been triggered by something I was trying to do in git that didn't work. My bad. I think it's ok now.

I left a comment on https://github.com/godotengine/godot/pull/105825 (the commit I conflicted with). I think that commit missed some of the code since mine had to modify two locations.

KoBeWi commented:
For reference, comparison of old and new behavior:

https://github.com/user-attachments/assets/c1519b95-f860-4e8b-ba85-c321dce6abf4

https://github.com/user-attachments/assets/1873f8fc-0c77-46c6-bc40-3886bd8f515a

Before this change, the camera is completely stuck in one spot.
After this change, it seems to be snapping to anchor points? Somewhat weird, but it's better probably.

KoBeWi commented on scene/2d/camera_2d.cpp:
```diff
@@ -219,19 +219,31 @@ Transform2D Camera2D::get_camera_transform() {
-			if (screen_rect.position.y < limit[SIDE_TOP]) {
+				if (screen_rect.position.y < limit[SIDE_TOP]) {
+					// Split the limit difference vertically.
+					camera_pos.y -= (screen_rect.position.y - limit[SIDE_TOP] + screen_rect.position.y + screen_rect.size.y - limit[SIDE_BOTTOM]) / 2;
```
Same as above.

ogapo committed:
Fix for camera limits
This commit makes 2D camera limits have a well defined behavior for cases when the limits are smaller than the screen rect.

Presently the RIGHT and TOP limit take prescidence simply because they are applied second. This change adjusts behavior to split the difference in cases where both the LEFT/RIGHT or TOP/BOTTOM limits would both apply.

ogapo commented:
Ok, I changed the if-else chain again. Expected behavior now is it should just lock to the center (rather than the top right corner) any time the limit bounds are smaller than the screen.

Repiteo commented:
Thanks!

kleiders3010 commented:
This change is very bothering, it changed my functional camera to be non-functional, by making it snap left and right. This is specially weird, since the camera limit is the same size as the screen rect

Why would the camera snapping from left to right be a more desired behaviour than it being static?

First half of the video is 4.4.1, second half is 4.5beta

https://github.com/user-attachments/assets/70c9a8bf-310f-46ae-af3b-268ecee2dedb

kleiders3010 commented:
To add onto this, setting Limit -> Smoothed to TRUE makes it behave like in 4.4.1, which makes it feel more like unintended behaviour
